src/pretix/base/services/orders.py

```diff
@@ -2619,14 +2619,34 @@ def _payment_fee_diff(self):
     def _reissue_invoice(self):
         i = self.order.invoices.filter(is_cancellation=False).last()
         if self.reissue_invoice and self._invoice_dirty:
-            if i and not i.refered.exists():
-                self._invoices.append(generate_cancellation(i))
-            if invoice_qualified(self.order) and \
-                (i or
-                 self.event.settings.invoice_generate == 'True' or (
-                     self.open_payment is not None and self.event.settings.invoice_generate == 'paid' and
-                     self.open_payment.payment_provider.requires_invoice_immediately)):
-                self._invoices.append(generate_invoice(self.order))
+            order_now_qualified = invoice_qualified(self.order)
+            invoice_should_be_generated = (
+                self.event.settings.invoice_generate == "True" or (
+                    self.event.settings.invoice_generate == "paid" and
+                    self.open_payment is not None and
+                    self.open_payment.payment_provider.requires_invoice_immediately
+                ) or (
+                    self.event.settings.invoice_generate == "paid" and
+                    self.order.status == Order.STATUS_PAID
+                ) or (
+                    # Backwards-compatible behaviour
+                    self.event.settings.invoice_generate not in ("True", "paid") and
+                    i and
+                    not i.refered.exists()
+                )
+            )
+
+            if order_now_qualified:
+                if invoice_should_be_generated:
+                    if i and not i.refered.exists():
```

Comment:
I am not sure, the code comments read `Invoice.refered` as a general way to refer invoices. Shouldn’t this be explicitly about refered.is_cancelation, ie.:
```suggestion
                    if i and not i.canceled:
```
Otherwise this code IMHO assumes that every refered is a cancellation – might be the case in reality though as it was used this way in older code (e.g. line 348 or 2622) …